model/subscriber.go

```diff
@@ -64,3 +70,55 @@ func (s *Subscriber) Copy(dst datastore.Entity) (datastore.Entity, error) {
 func (s *Subscriber) GetCache() datastore.Cache {
 	return nil
 }
+
+// CreateSubscriber creates a subscriber, or returns an error if a subscriber with the
+// given ID already exists.
+func CreateSubscriber(ctx context.Context, store datastore.Store, s *Subscriber) error {
+	key := store.IDKey(typeSubscriber, s.ID)
+	return store.Create(ctx, key, s)
+}
+
+// GetSubscriberByEmail returns the subscriber with the given email if it exists.
+func GetSubscriberByEmail(ctx context.Context, store datastore.Store, email string) (*Subscriber, error) {
+	q := store.NewQuery(typeSubscriber, false)
+	q.FilterField("Email", "=", email)
+	var subs []Subscriber
+	_, err := store.GetAll(ctx, q, &subs)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get all subscribers: %w", err)
+	}
+
+	if len(subs) == 0 {
+		return nil, datastore.ErrNoSuchEntity
+	}
+
+	if len(subs) > 1 {
+		return nil, errDuplicateSubscriberEmails
+	}
+
+	return &subs[0], err
+}
+
+// GetSubscriber gets the subscriber with the given ID.
+func GetSubscriber(ctx context.Context, store datastore.Store, id int64) (*Subscriber, error) {
+	key := store.IDKey(typeSubscriber, id)
+	sub := &Subscriber{}
+	err := store.Get(ctx, key, sub)
+	return sub, err
+}
+
+// NewSubscriberID returns a new unique in64 ID which does not already exist in
+// the datastore.
+func NewSubscriberID(ctx context.Context, store datastore.Store) int64 {
+	for {
+		// Create a random 31-bit number that is at least 10 digits.
+		id := rand.Int63n((1<<31)-1000000000) + 1000000000
+
+		// Check if the ID is unique.
+		_, err := GetSubscriber(ctx, store, id)
```

Comment:
As mentioned above, this isn't concurrency safe.

Reply:
You want to call `CreateSubscriber` until it is successful, as we do for sites, e.g.:

```
	for {
		// Create a random 31-bit number that is at least 10 digits.
		skey = rand.Int63n((1<<31)-1000000000) + 1000000000
		site := model.Site{Skey: skey, Name: sn, Public: false, Enabled: true}
		err := model.CreateSite(ctx, settingsStore, &site)
		if err == nil {
			break
		}
		if err != datastore.ErrEntityExists {
			return fmt.Errorf("cannot create site: %w", err)
		}
	}
```

Note that his still not concurrency safe in that it could result in two sites with the same name but different site keys. One site would be a zombie site that never gets used, though.

Reply:
I have now addressed this by having the Create function optionally generate a unique ID for the new subscriber through this method.